pgdb/v1/schema.go

```diff
@@ -183,3 +227,60 @@ func Migrations(ctx context.Context, db sqlScanner, msg DBReflectMessage) ([]str
 	}
 	return rv, nil
 }
+
+func sha256String(input string) string {
+	h := sha256.New()
+	_, _ = h.Write([]byte(input))
+	return hex.EncodeToString(h.Sum(nil))
+}
+
+// This appends a new hash of the tenantId to the end of the table name.
+// If the length is over PGs max table name length then we want to preserve the original hash.
+// So we split the table name and create the new ending hashes like so 0565c036_12345678
+// Then if its over limit we but down the table name in order to append the new ending hashes.
+// (this isnt over 63 but for example)
+// pb_pasta_ingredient_models_food_v1_0565c036 -> pb_pasta_ingredient_models_0565c036_12345678.
+func createPartitionTableName(tableName string, tenantId string) string {
+	const pgMaxTableNameLen = 63
+	tenantHash := sha256String(tenantId)[0:8]
+	newName := tableName + "_" + tenantHash
+	if len(newName) < pgMaxTableNameLen {
+		return newName
+	}
+	tableSplit := strings.Split(tableName, "_")
+	originalHash := tableSplit[len(tableSplit)-1]
+	nameWithoutHash := strings.Join(tableSplit[0:len(tableSplit)-1], "_")
+	combinedHashes := originalHash + "_" + tenantHash
+	nameWithoutHash = nameWithoutHash[0 : pgMaxTableNameLen-len(combinedHashes)]
+	return nameWithoutHash + "_" + combinedHashes
+}
+
+// This will be passed in in C1.
+type TenantIteratorFunc func(ctx context.Context) (string, error)
+type SchemaUpdateFunc func(ctx context.Context, tenantId string, schema string) error
+
+func TenantPartitionsUpdate(ctx context.Context, msg DBReflectMessage, iteratorFunc TenantIteratorFunc, updateFunc SchemaUpdateFunc) {
+	tableName := msg.DBReflect().Descriptor().TableName()
```

Comment:
Right so if someone sets this partition option on an existing model should we maybe prevent a possible problem by checking that a non-partitioned version of the table exists until we explicitly support altering an existing table to have it be partitioned. That is make it maybe a noop that logs "skipping created partitioned table for existing table" or dpanic or error or something?

Reply:
Right good point ill add a check

Reply:
There is also an error thrown in postgres if you try to make a partition for a non-partitioned table
```
 ERROR: "pb_pasta_models_food_v1_29fd1107" is not partitioned (SQLSTATE 42P17)
 ```

Reply:
I did a quick test to override creating a partition and seems to work by skipping
```
=== RUN   TestSchemaFoodPasta
Skipping creating partitioned tables for a table that is not marked as a parent (partitioned) table: pb_pasta_models_food_v1_29fd1107    
food_test.go:208: 
                Error Trace:    /home/mitscott/src/protoc-gen-pgdb/example/models/food/v1/food_test.go:208
                                                        /home/mitscott/src/protoc-gen-pgdb/example/models/food/v1/food_test.go:128
                Error:          Not equal: 
                                expected: 3
                                actual  : 0
                Test:           TestSchemaFoodPasta
                Messages:       Should have one sub-partition table per fake tenant
--- FAIL: TestSchemaFoodPasta (0.41s)
```

Reply:
https://github.com/ductone/protoc-gen-pgdb/pull/63/commits/16b385bcbf5fa646916852942da7c62925d6c762